application/src/main/java/org/thingsboard/server/service/gateway_device/DefaultGatewayDeviceStateService.java

```diff
@@ -0,0 +1,115 @@
+
+    private final RelationService relationService;
+    private final DeviceService deviceService;
+
+    @Lazy
+    @Autowired
+    private TbCoreDeviceRpcService deviceRpcService;
+
+    @Override
+    public void update(Device device, Device oldDevice) {
+        Device gatewayDevice = findGatewayDeviceByRelationFromDevice(device);
+        if (gatewayDevice != null) {
+            ObjectNode renamedDeviceNode = JacksonUtil.newObjectNode();
+            renamedDeviceNode.put(device.getName(), oldDevice.getName());
+            ToDeviceRpcRequest rpcRequest = formDeviceToGatewayRPCRequest(gatewayDevice, renamedDeviceNode, DEVICE_RENAMED_METHOD_NAME);
+            deviceRpcService.processRestApiRpcRequest(rpcRequest, fromDeviceRpcResponse -> {
+                log.trace("Device renamed RPC with id: [{}] processed to gateway device with id: [{}], old device name: [{}], new device name: [{}]",
+                        rpcRequest.getId(), gatewayDevice.getId(), oldDevice.getName(), device.getName());
+            }, null);
+        }
+    }
+
+    @Override
+    public void delete(Device device) {
+        Device gatewayDevice = findGatewayDeviceByRelationFromDevice(device);
+        if (gatewayDevice != null) {
+            TextNode deletedDeviceNode = new TextNode(device.getName());
+            ToDeviceRpcRequest rpcRequest = formDeviceToGatewayRPCRequest(gatewayDevice, deletedDeviceNode, DEVICE_DELETED_METHOD_NAME);
+            deviceRpcService.processRestApiRpcRequest(rpcRequest, fromDeviceRpcResponse -> {
+                log.trace("Device deleted RPC with id: [{}] processed to gateway device with id: [{}], deleted device name: [{}]",
+                        rpcRequest.getId(), gatewayDevice.getId(), device.getName());
+            }, null);
+        }
+    }
+
+    private ToDeviceRpcRequest formDeviceToGatewayRPCRequest(Device gatewayDevice, JsonNode deviceDataNode, String method) {
+        ToDeviceRpcRequestBody body = new ToDeviceRpcRequestBody(method, JacksonUtil.toString(deviceDataNode));
+        long expTime = System.currentTimeMillis() + Math.max(minTimeout, defaultTimeout);
+        UUID rpcRequestUUID = UUID.randomUUID();
+        return new ToDeviceRpcRequest(rpcRequestUUID,
+                gatewayDevice.getTenantId(),
+                gatewayDevice.getId(),
+                true,
+                expTime,
+                body,
+                true,
+                null,
+                null
+        );
+    }
+
+    private Device findGatewayDeviceByRelationFromDevice(Device device) {
+        List<EntityRelation> relationToGatewayList = relationService.findByFromAndType(TenantId.SYS_TENANT_ID, device.getId(), DataConstants.LAST_CONNECTED_GATEWAY, RelationTypeGroup.COMMON);
+        if (!relationToGatewayList.isEmpty()) {
+            EntityRelation relationToGateway = relationToGatewayList.get(0);
+            return deviceService.findDeviceById(device.getTenantId(), (DeviceId) relationToGateway.getTo());
```

Comment:
You don't need the device itself. you only need the device id. The Tenant id is always the same.